### PR TITLE
8343493: Perform module checks during MetaspaceShared::map_archives()

### PR DESCRIPTION
**Pull request by iklam**
Currently, `Modules::serialize()` and `Modules::serialize_addmods_names()` store/retrieve the main module name, as well as the set of modules specified by `--add-modules`. These functions are called from `MetaspaceShared::serialize()`, *after* the JVM has decided to use the CDS archive.

However, with JEP 483 ([JDK-8315737](https://bugs.openjdk.org/browse/JDK-8315737)), we need the module information *before* deciding whether to use the CDS archive. For example, if `-XX:+AOTClassLinking` is specified during the CDS archive creation, the archive cannot be used if we specified an incompatible main module at runtime.

This RFE added a new function, `MetaspaceShared::early_serialize()`, which is called from `MetaspaceShared::map_archives()`. `Modules::serialize()` and `Modules::serialize_addmods_names()` are called within this context so that we can use the module information to decide whether to use the CDS archive or not.

I also added some comments about what "serialize" does, since it's rather non-obvious.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343493](https://bugs.openjdk.org/browse/JDK-8343493): Perform module checks during MetaspaceShared::map_archives() (**Enhancement** - P4)

### Reviewers
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - **Reviewer**) Review applies to [83914209](https://git.openjdk.org/jdk/pull/21890/files/8391420924dbe18f93976b20b82001f887ae1090)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21890/head:pull/21890` \
`$ git checkout pull/21890`

Update a local copy of the PR: \
`$ git checkout pull/21890` \
`$ git pull https://git.openjdk.org/jdk.git pull/21890/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21890`

View PR using the GUI difftool: \
`$ git pr show -t 21890`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21890.diff">https://git.openjdk.org/jdk/pull/21890.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21890#issuecomment-2456118743)
</details>

### CODE REVIEW

File: src/hotspot/share/cds/filemap.hpp [hunk 1]
```diff
@@ -194,6 +194,7 @@ class FileMapHeader: private CDSFileMapHeaderBase {
   bool    _compressed_class_ptrs;                 // save the flag UseCompressedClassPointers
   bool    _use_secondary_supers_table;            // save the flag UseSecondarySupersTable
   size_t  _cloned_vtables_offset;                 // The address of the first cloned vtable
+  size_t  _early_serialized_data_offset;          // Data accessed using {ReadClosure,WriteClosure}::serialize()
```
Comment (calvinccheung) [inline src/hotspot/share/cds/filemap.hpp:197]: Should we print this field in `FileMapHeader::print()`?

**Review by calvinccheung — COMMENTED** [review]
Looks good overall. One comment:

File: src/hotspot/share/cds/filemap.hpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/filemap.hpp:197]: I added the print out as you suggested.

File: src/hotspot/share/cds/metaspaceShared.cpp [hunk 2]
```diff
@@ -359,8 +359,43 @@ void MetaspaceShared::read_extra_data(JavaThread* current, const char* filename)
...
+//     At runtime, these two variables are loaded from the CDS archive.
+//     In addition, the pointer is relocated as necessary.
+//
+// Some of the xxx::serialize() functions may have side effects and assume that
+// the archive is already mapped. For example, SymbolTable::serialize_shared_table_header()
+// unconditionally makes the set of archived symbols available. Therefore, we put most
+// of these xxx::serialize() functions inside MetaspaceShared::serialize(), which
+// is called AFTER we made the decision to map the archive.
+//
+// However, some of the "seralized" data are used to decide whether an archive should
```
Comment (matias9927) [inline src/hotspot/share/cds/metaspaceShared.cpp:386]: Typo: seralized -> serialized

**Review by calvinccheung — APPROVED** [review]
Thanks for the update.

**Review by calvinccheung — APPROVED** [review]

**Comment by iklam** [comment]
Thanks @calvinccheung and @matias9927 for review.
/integrate

**Pull request closed without merging**